[scroll-animations] Implement timeline-scope

nmoucht opened the pull request:
#### 28e4fdec7bdd67ab22f5b205e0a57305150d88c0
<pre>
[scroll-animations] Implement timeline-scope
<a href="https://bugs.webkit.org/show_bug.cgi?id=282072">https://bugs.webkit.org/show_bug.cgi?id=282072</a>
<a href="https://rdar.apple.com/138597234">rdar://138597234</a>

Reviewed by Antoine Quint.

Scope named timeline lookups to particular subtree of declared element&apos;s subtree and
implement the timeline-scope property.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-named-scroll-progress-timeline.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-in-container-query-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-multi-pass.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-range-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-lookup-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-range-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-used-values-expected.txt:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::originatingElement):
(WebCore::originatingElementIncludingTimelineScope):
(WebCore::originatingElementExcludingTimelineScope):
(WebCore::AnimationTimelinesController::relatedTimelineScopeElements):
(WebCore::determineTreeOrder):
(WebCore::determineTimelineForElement):
(WebCore::AnimationTimelinesController::updateTimelineForTimelineScope):
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
(WebCore::AnimationTimelinesController::attachPendingOperations):
(WebCore::AnimationTimelinesController::registerNamedViewTimeline):
(WebCore::AnimationTimelinesController::unregisterNamedTimeline):
(WebCore::AnimationTimelinesController::setTimelineForName):
(WebCore::updateTimelinesForTimelineScope):
(WebCore::AnimationTimelinesController::updateNamedTimelineMapForTimelineScope):
(WebCore::AnimationTimelinesController::timelineForName const): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncPropertiesWithBackingAnimation):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::setTimelineScopeElement):
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::timelineScopeDeclaredElement const):
(WebCore::ScrollTimeline::clearTimelineScopeDeclaredElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Canonical link: <a href="https://commits.webkit.org/286771@main">https://commits.webkit.org/286771@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcea93e86ff1968a3ec40cfecdd8d0b99a7c2b6d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76999 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56034 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29914 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81547 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28278 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79116 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4330 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60343 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18416 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80066 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50287 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66091 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40654 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47689 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23588 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26604 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68809 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23917 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82982 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2937 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68615 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4534 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66064 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67866 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11852 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9932 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11919 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4325 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7141 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4345 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7780 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6104 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -310,6 +311,62 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+{
+    RefPtr element = &matchElement;
+    while (element) {
+        auto i = elementAncestors.findIf([&] (const Ref<ScrollTimeline>& entry) {
```
You can use `auto` instead of `Ref<ScrollTimeline>`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -310,6 +311,62 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
         m_cachedCurrentTime = std::nullopt;
 }
 
+static Element* originatingElement(Ref<ScrollTimeline> timeline)
```
`Ref<>&`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -310,6 +311,62 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    return timeline->source();
+}
+
+static RefPtr<ScrollTimeline> determineTreeOrder(Vector<Ref<ScrollTimeline>> elementAncestors, const Element& matchElement)
```
I'm not sure returning `RefPtr<>` is the right thing to do. I _think_ the guidance is to let call sites worry about creating a `RefPtr<>` and we could return simply a `const ScrollTimeline*`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -310,6 +311,62 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    return nullptr;
+}
+
+static RefPtr<ScrollTimeline> determineTimelineForElement(Vector<Ref<ScrollTimeline>> entries, const Element& matchElement)
```
Same comment as above about method signature.

graouts replied:
Maybe use `timelines` rather than `entries`? The second parameter probably could be just `element`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -310,6 +311,62 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    // If multiple elements have declared the same timeline name, the matching timeline is the one declared on the nearest element in tree order.
+    // In case of a name conflict on the same element, names declared later in the naming property (scroll-timeline-name, view-timeline-name) take
+    // precedence, and scroll progress timelines take precedence over view progress timelines.
+    Vector<Ref<ScrollTimeline>> elementAncestors;
```
I would expect something called `elementAncestors` to contain elements, not timelines.

graouts replied:
Could this be a static method?

nmoucht replied:
Changed to ancestorTimelines.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -310,6 +311,62 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    // precedence, and scroll progress timelines take precedence over view progress timelines.
+    Vector<Ref<ScrollTimeline>> elementAncestors;
+    for (auto& entry : entries) {
+        if (originatingElement(entry) ==  &matchElement || matchElement.isDescendantOf(originatingElement(entry)))
```
Extra space before `&matchElement`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -310,6 +311,62 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    if (entries.size() == 1) {
+        auto& entry = entries.first();
+        for (auto& element : entry->timelineScopeDeclaredElements()) {
+            if (originatingElement(entry) ==  &element || element.isDescendantOf(originatingElement(entry)))
```
We might need a smart pointer for `element`.

nmoucht replied:
No longer relevant.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -364,26 +421,53 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
+    case TimelineScope::Type::All:
+        for (auto& entry : m_nameToTimelineMap) {
+            updateTimelineForTimelineScope(entry.value, element);
+        }
```
No need for curly braces for a single-line statement.

nmoucht replied:
No longer relevant.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -364,26 +421,53 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
+                continue;
+
+            auto& timelines = it->value;
+            updateTimelineForTimelineScope(timelines, element);
```
You can probably forgo the use of this variable and just pass `it->value` straight to `updateTimelineForTimelineScope()`.

nmoucht replied:
Fixed.

graouts commented on LayoutTests/TestExpectations:
```diff
@@ -7401,6 +7401,8 @@ imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-range-update
 imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-range-update.html [ Skip ]
 imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-subject-bounds-update.html [ Skip ]
 
+webkit.org/b/283149 imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-inset-animation.html [ Skip ]
```
Out of interest, where does this test crash now?

nmoucht replied:
The same place. I'll make a follow up after this lands.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,111 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    return timeline->source();
+}
+
+static Element* originatingElementIncludingTimelineScope(Ref<ScrollTimeline> timeline)
```
`Ref<>&`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,111 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    return originatingElement(timeline);
+}
+
+static Element* originatingElementExcludingTimelineScope(Ref<ScrollTimeline> timeline)
```
`Ref<>&`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,111 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+
+Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>> AnimationTimelinesController::relatedTimelineScopeElements(const AtomString& name)
+{
+    Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>> ret;
```
Can we find a more descriptive name than `ret` here?

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,111 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    return ret;
+}
+
+RefPtr<ScrollTimeline> AnimationTimelinesController::determineTreeOrder(Vector<Ref<ScrollTimeline>> elementAncestors, const Element& matchElement, Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>> timelineScopeElements)
```
`Vector<Ref<ScrollTimeline>>&`, maybe even `const`?

graouts replied:
Why do we return a `RefPtr<ScrollTimeline>`? I _think_ the guideline for memory safety is for call sites to determine whether they need a `RefPtr` and we should return a raw pointer? Or otherwise return a reference to an existing `RefPtr`?

graouts replied:
Could this be a static method?

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,111 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+                matchedTimelines.append(timeline);
+        }
+        if (!matchedTimelines.isEmpty()) {
+            // Conflict due to timeline-scope
```
What does this mean?

nmoucht replied:
Moved this comment to the return nullptr case. This is for the case where there are multiple elements that are relevant for the timeline-scope, which we can't resolve.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,111 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    return nullptr;
+}
+
+RefPtr<ScrollTimeline> AnimationTimelinesController::determineTimelineForElement(Vector<Ref<ScrollTimeline>> entries, const Element& matchElement, Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>> timelineScopeElements)
```
Same comment about the return value being `RefPtr` as for `determineTreeOrder()`.

graouts replied:
Can both provided vectors be const references?

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,111 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+void AnimationTimelinesController::updateTimelineForTimelineScope(Ref<ScrollTimeline> timeline, const AtomString& name)
+{
+    WeakHashSet<Element, WeakPtrImplWithEventTargetData> matchedTimelineScopeElements;
+    auto* timelineElement = originatingElementExcludingTimelineScope(timeline);
```
`RefPtr`.

graouts replied:
I'd move this further down where we actually start using that variable.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -367,27 +482,91 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
+        return originatingElement(entry) == &element;
     });
+
+    if (i != notFound) {
```
Can we return early if `i == notFound` and `ASSERT(m_pendingAttachOperations.isEmpty())`?

nmoucht replied:
I don't think we can do that assert. Added the return early.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -367,27 +482,91 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
+        timelines.remove(i);
+    }
+
     if (timelines.isEmpty())
```
This can move in the `if` block above I believe (unless we do return early).

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -367,27 +482,91 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
-    if (it == m_nameToTimelineMap.end())
-        return nullptr;
+    if (it == m_nameToTimelineMap.end()) {
+        m_pendingAttachOperations.append(TimelineMapAttachOperation { element, name, animation });
```
Presumably we can remove `TimelineMapAttachOperation` here.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -367,27 +482,91 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
-    if (timelines.isEmpty())
-        return nullptr;
-    return timelines.first().ptr();
+    m_pendingAttachOperations.append(TimelineMapAttachOperation { element, name, animation });
```
Ditto about dropping `TimelineMapAttachOperation`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -367,27 +482,91 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
+                continue;
+
+            auto& timelines = it->value;
+            updateTimelinesForTimelineScope(timelines, element);
```
```suggestion
            if (it != m_nameToTimelineMap.end())
                updateTimelinesForTimelineScope(it->value, element);
```

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,112 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    return timelineScopeElements;
+}
+
+static ScrollTimeline* determineTreeOrder(Vector<Ref<ScrollTimeline>> ancestorTimelines, const Element& matchElement, Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>> timelineScopeElements)
```
The provided vectors should be const references.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -314,13 +316,112 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+            if (timelineScopeElements.contains(element.get())) {
+                if (matchedTimelines.size() == 1)
+                    return matchedTimelines.first().ptr();
+                // Conflict due to timeline-scope
```
Still not sure what this comment means.

nmoucht replied:
This means that multiple elements could be associated due to timeline-scope, which results in an inactive timeline.

graouts replied:
This would have been a better comment to put in the commit.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -367,27 +483,88 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
+
+    if (i == notFound)
+        return;
+    auto& timeline = timelines.at(i);
```
I think an empty line between those two statements would help readability.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -367,27 +483,88 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
-    return timelines.first().ptr();
+    m_pendingAttachOperations.append({ element, name, animation });
+
+    return;
```
Get rid of that.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -367,27 +483,88 @@ void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& nam
+static void updateTimelinesForTimelineScope(Vector<Ref<ScrollTimeline>> entries, const Element& element)
+{
+    for (auto& entry : entries) {
+        if (auto entryElement = originatingElementExcludingTimelineScope(entry)) {
```
`RefPtr`

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-animations] Implement timeline-scope
https://bugs.webkit.org/show_bug.cgi?id=282072
rdar://138597234

Reviewed by Antoine Quint.

Scope named timeline lookups to particular subtree of declared element's subtree and
implement the timeline-scope property.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-named-scroll-progress-timeline.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-in-container-query-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-multi-pass.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-range-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-scope-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-lookup-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-range-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-used-values-expected.txt:
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::originatingElement):
(WebCore::originatingElementIncludingTimelineScope):
(WebCore::originatingElementExcludingTimelineScope):
(WebCore::AnimationTimelinesController::relatedTimelineScopeElements):
(WebCore::determineTreeOrder):
(WebCore::determineTimelineForElement):
(WebCore::AnimationTimelinesController::updateTimelineForTimelineScope):
(WebCore::AnimationTimelinesController::registerNamedScrollTimeline):
(WebCore::AnimationTimelinesController::attachPendingOperations):
(WebCore::AnimationTimelinesController::registerNamedViewTimeline):
(WebCore::AnimationTimelinesController::unregisterNamedTimeline):
(WebCore::AnimationTimelinesController::setTimelineForName):
(WebCore::updateTimelinesForTimelineScope):
(WebCore::AnimationTimelinesController::updateNamedTimelineMapForTimelineScope):
(WebCore::AnimationTimelinesController::timelineForName const): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncPropertiesWithBackingAnimation):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::setTimelineScopeElement):
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::timelineScopeDeclaredElement const):
(WebCore::ScrollTimeline::clearTimelineScopeDeclaredElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Canonical link: https://commits.webkit.org/286771@main
